### PR TITLE
Fix/add erc721 safe mint and safe transfer from #1736

### LINKED ISSUE #1736: Add ERC721 _safeMint and _safeTransferFrom
`ERC721` has `safeTransferFrom` that forces a receiver contract to implement the `onERC721Received` function. Just like our other ERC implementations, we should have an internal `_safeTransferFrom` that does the same thing without requiring anything of `msg.sender`. Note that we already have an internal `_transferFrom`.

Similarly, we should add `_safeMint` as an alternative to `_mint`, with the addition that it also call `onERC721Received` on a contract receiver.

***

Original issue:

> `ERC721` transfers are required to call `onERC721Received` if the receiver is a contract. I don't know if this is specified in the EIP, but we might have to do the same in `_mint`. Currently we are not.
>
> This needs some discussion because it would be a breaking change. Additionally, it may not be obvious that `_mint` will call an external contract, and as always this opens up the possibility for reentrancy.
>
> Thanks @alcuadrado for pointing this out.

### PR DESCRIPTION
**Pull request by balajipachai**
Fixes #1736

Created this PR in reference to #1736 
Added _safeMint() & _safeTransferFrom() in ERC721.sol
Tested the ERC721.test.js, all test cases are executing as expected.
However in coverage report, for ERC721.sol, percentage of branch covered is 97.06%, the else branch of require() in _safeMint() is not covered.

Just go through the implementation of _safeMint() & _safeTransferFrom(), let me know, in case of any changes.
Once these functionalities are freezed, then I will add the **test case** for covering the else branch of  require() in _safeMint()

Thanks!!!

### CODE REVIEW

File: contracts/token/ERC721/ERC721.sol [hunk 1]
```diff
@@ -165,7 +165,7 @@ contract ERC721 is ERC165, IERC721 {
      */
     function safeTransferFrom(address from, address to, uint256 tokenId, bytes memory _data) public {
         transferFrom(from, to, tokenId);
-        require(_checkOnERC721Received(from, to, tokenId, _data));
+        _safeTransferFrom(from, to, tokenId, _data);
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721.sol]: I think you may have misinterpreted @frangio's suggestion: the idea was for `_safeTransferFrom` to call both `_transferFrom` and `__checkOnERC721Received`.

**Comment by nventuro** [comment]
Hello @balajipachai, thank you for contributing!

I think you may have misinterpreted @frangio's suggestion for this feature: the idea is for `_safeTransferFrom` to call both `_transferFrom` and `_checkOnERC721Received` (same with `_safeMint`), as opposed to simply doing the recipient check.

Something like

```
function safeTransferFrom {
   require(_isApprovedOrOwner(msg.sender, tokenId));
  _safeTransferFrom(msg.sender);
}

function _safeTransferFrom {
  _transferFrom(from, to, tokenId);
  _checkOnERC721Received
}
```

**Comment by balajipachai** [comment]
> Hello @balajipachai, thank you for contributing!
> 
> I think you may have misinterpreted @frangio's suggestion for this feature: the idea is for `_safeTransferFrom` to call both `_transferFrom` and `_checkOnERC721Received` (same with `_safeMint`), as opposed to simply doing the recipient check.
> 
> Something like
> 
> ```
> function safeTransferFrom {
>    require(_isApprovedOrOwner(msg.sender, tokenId));
>   _safeTransferFrom(msg.sender);
> }
> 
> function _safeTransferFrom {
>   _transferFrom(from, to, tokenId);
>   _checkOnERC721Received
> }
> ```

Adding _transferFrom() in _safeMint() where the **from** address is **address(0)**, this causes the ERC721.test.js to fails at the initial level itself.
In _safeMint(), thus, I have commented, _transferFrom() line of code, you can uncomment it and notice that the test cases start failing.
Provided this, I think, there is no need to add _transferFrom() inside the _safeMint() function.

Awaiting your inputs.

**Comment by nventuro** [comment]
You're right, my bad, I thought `_mint` was written in terms of `_transferFrom`, but it isn't. We'd have to call `_mint` from `_safeMint` then.

**Comment by balajipachai** [comment]
> You're right, my bad, I thought `_mint` was written in terms of `_transferFrom`, but it isn't. We'd have to call `_mint` from `_safeMint` then.

@nventuro 
I think instead of calling `_mint` from `_safeMint` the correct way should be calling `_safeMint()` from `_mint`
**Why?**
We find the occurrence of `_mint` in **ERC721Enumerable.sol, ERC721MetadataMintable.sol, ERC721Mintable**

> If we call `_mint` from `_safeMint`, it implies, changing `_mint` to `_safeMint` in the above files
> `_safeMint` takes 4 parameters **address from, address to, uint256 tokenId, bytes memory _data**, this implies the addition of **from, _data** in the `_safeMint` call instead of `_mint` call, which I think is cumbersome and tedious and leads to changes in the .sol files and also rewriting of the test files which is not required as per my understanding.

```
Thus, we should call `_safeMint` from `_mint` and not `_mint` from `_safeMint`.
```

Hope it helps !!!!!!!!!

**Comment by frangio** [comment]
@balajipachai I'm having some trouble understanding exactly what you're trying to say. 

>If we call `_mint` from `_safeMint`, it implies, changing `_mint` to `_safeMint` in the above files

How does it imply that? `_safeMint` is a new function, an alternative to `_mint`. `_mint` should not change anywhere.

**Comment by frangio** [comment]
I edited the original issue slightly to be more clear on this:
>Similarly, we should add `_safeMint` as an alternative to `_mint`, with the addition that it also call `onERC721Received` on a contract receiver.

**Comment by pachaibalaji** [comment]
> @balajipachai I'm having some trouble understanding exactly what you're trying to say.
> 
> > If we call `_mint` from `_safeMint`, it implies, changing `_mint` to `_safeMint` in the above files
> 
> How does it imply that? `_safeMint` is a new function, an alternative to `_mint`. `_mint` should not change anywhere.

Ok, I thought, _safeMint is a replacement to _mint, my bad !!!

**Comment by balajipachai** [comment]
> @balajipachai I'm having some trouble understanding exactly what you're trying to say.
> 
> > If we call `_mint` from `_safeMint`, it implies, changing `_mint` to `_safeMint` in the above files
> 
> How does it imply that? `_safeMint` is a new function, an alternative to `_mint`. `_mint` should not change anywhere.

OK, I thought, `_safeMint` is a replacement to `_mint`, now things are clear, thanks !!!

File: contracts/token/ERC721/ERC721.sol [hunk 2]
```diff
@@ -140,10 +140,8 @@ contract ERC721 is ERC165, IERC721 {
      * @param tokenId uint256 ID of the token to be transferred
      */
     function transferFrom(address from, address to, uint256 tokenId) public {
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: Please revert this function to the way it was before.

File: contracts/token/ERC721/ERC721.sol [hunk 3]
```diff
@@ -174,8 +172,8 @@ contract ERC721 is ERC165, IERC721 {
      * @param _data bytes data to send along with a safe transfer check
      */
     function safeTransferFrom(address from, address to, uint256 tokenId, bytes memory _data) public {
-        transferFrom(from, to, tokenId);
-        require(_checkOnERC721Received(from, to, tokenId, _data), "ERC721: transfer to non ERC721Receiver implementer");
+        require(_isApprovedOrOwner(msg.sender, tokenId));
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: We should add a revert reason here. I think you can just copy the one in `transferFrom`, right @nventuro?

File: contracts/token/ERC721/ERC721.sol [hunk 4]
```diff
@@ -208,8 +206,8 @@ contract ERC721 is ERC165, IERC721 {
      * @param tokenId uint256 ID of the token to be minted
      */
     function _mint(address to, uint256 tokenId) internal {
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: Please revert this function to the way it was before.

File: contracts/token/ERC721/ERC721.sol [hunk 5]
```diff
@@ -284,6 +282,28 @@ contract ERC721 is ERC165, IERC721 {
         bytes4 retval = IERC721Receiver(to).onERC721Received(msg.sender, from, tokenId, _data);
         return (retval == _ERC721_RECEIVED);
     }
+    /**
+    * @dev Internal function that safely transfers the ownership of a given token ID to another address
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]:
```suggestion
     * @dev Internal function that safely transfers the ownership of a given token ID to another address
```

File: contracts/token/ERC721/ERC721.sol [hunk 6]
```diff
@@ -284,6 +282,28 @@ contract ERC721 is ERC165, IERC721 {
...
+     * @param from current owner of the token
+     * @param to address to receive the ownership of the given token ID
+     * @param tokenId uint256 ID of the token to be transferred
+     * @param _data bytes data to send along with a safe transfer check
+     */
+    function _safeTransferFrom(address from, address to, uint256 tokenId, bytes memory _data) internal {
+        require(_checkOnERC721Received(from, to, tokenId, _data));
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: We should add a revert reason here, as well as in the equivalent `require` in `_safeMint`. You can copy the one that was in `safeTransferFrom`.

**Pull request closed without merging**